pkg/storage/mvcc.go

```diff
@@ -2218,6 +2274,211 @@ func recordIteratorStats(traceSpan *tracing.Span, iteratorStats IteratorStats) {
+			deleteRangeStart = nil
+			batchBytes += int64(encodedMVCCKeyLength(MVCCKey{Key: nonMatch})) // account for end key
+			return err
+		}
+
+		if deleteBufIdx > 0 {
+			iter := newMVCCIterator(rw, false, IterOptions{Prefix: true})
+			defer iter.Close()
+			for i := 0; i < deleteBufIdx; i++ {
+				err := mvccPutInternal(
+					ctx, rw, iter, ms, deleteBuf[i], writeTimestamp, nil, nil, putBuf, nil)
+				if err != nil {
+					return err
+				}
+			}
+			deleteBufIdx = 0
+			deleteBufBytes = 0
+		}
+		return nil
+	}
+
+	revert := func(k roachpb.Key, v []byte) (*roachpb.Key, error) {
+		// For non-deletions, we have to flush any pending deletes first. This may also
+		// flush a range tombstone, which will add to batchBytes.
+		if len(v) > 0 {
+			if err := flushDeletes(k); err != nil {
+				return nil, err
+			}
+		}
+
+		// If the batch is full, return a resume key after flushing any deletes.
+		if batchSize >= maxBatchSize || batchBytes >= maxBatchBytes {
+			err := flushDeletes(k)
+			return &k, err
+		}
+		bytes := int64(encodedMVCCKeyLength(MVCCKey{Key: k, Timestamp: writeTimestamp}) + len(v))
+
+		if len(v) > 0 || len(deleteBuf) == 0 {
+			batchSize++
+			batchBytes += bytes
+			iter := newMVCCIterator(rw, false, IterOptions{Prefix: true})
+			defer iter.Close()
+			return nil, mvccPutInternal(ctx, rw, iter, ms, k, writeTimestamp, v, nil, putBuf, nil)
+
+		} else if len(deleteRangeStart) == 0 {
+			// We're currently buffering point deletions.
+			if deleteBufIdx < len(deleteBuf) {
+				deleteBuf[deleteBufIdx] = append(deleteBuf[deleteBufIdx][:0], k...)
+				deleteBufIdx++
+				deleteBufBytes += bytes
+				batchSize++
+				batchBytes += bytes
+			} else {
+				// Buffer is full -- switch to tracking the start of the range delete. We
+				// remove the buffered keys from the batch size, and instead only track
+				// the range key.
+				batchSize -= int64(deleteBufIdx) - 1 // -1 accounts for the range key
+				batchBytes -= deleteBufBytes -
+					int64(encodedMVCCKeyLength(MVCCKey{Key: deleteBuf[0], Timestamp: writeTimestamp}))
+				deleteRangeStart = deleteBuf[0]
+				deleteBufIdx = 0
+				deleteBufBytes = 0
+			}
+		}
+		return nil, nil
+	}
+
+	// We set up an ingremental iterator from the revert time to look for any
+	// changes that need to be reverted. However, we also need to inspect older
+	// values to e.g. find the value to revert to or make sure we don't drop range
+	// tombstones across them -- we do this by using the IgnoringTime() methods on
+	// the MVCCIncrementalIterator.
+	iter := NewMVCCIncrementalIterator(rw, MVCCIncrementalIterOptions{
+		EnableTimeBoundIteratorOptimization: true,
+		EndKey:                              endKey,
+		StartTime:                           revertTimestamp,
+		EndTime:                             writeTimestamp, // puts will error on any newer versions
+	})
+	defer iter.Close()
+
+	var revertKey roachpb.Key
+	var revertValue, revertValueFrom []byte
+	iter.SeekGE(MVCCKey{Key: startKey})
+	for {
```

Comment:
I found following mvcc iterator loops where we iterate versions and keys and feed it to "state machine" pretty hard to follow. Maybe the better idea would be to nest iteration of versions into iteration of keys and only track deletions inbetween?
We may leave it as a todo since we have some deadlines.